bin/cheritest/cheritest_vm.c

```diff
@@ -122,6 +123,144 @@ cheritest_vm_tag_shm_open_anon_private(const struct cheri_test *ctp __unused)
 	cheritest_success();
 }
 
+/*
+ * Test aliasing of SHM_ANON objects
+ */
+void
+cheritest_vm_tag_shm_open_anon_shared2x(const struct cheri_test *ctp __unused)
+{
+	void * __capability volatile * map2;
+	void * __capability c2;
+	int fd = CHERITEST_CHECK_SYSCALL(shm_open(SHM_ANON, O_RDWR, 0600));
+	CHERITEST_CHECK_SYSCALL(ftruncate(fd, getpagesize()));
+
+	map2 = CHERITEST_CHECK_SYSCALL(mmap(NULL, getpagesize(),
+		PROT_READ, MAP_SHARED, fd, 0));
+
+	/* Verify that no capability present */
+	c2 = *map2;
+	CHERITEST_VERIFY2(cheri_gettag(c2) == 0, "tag exists on first read");
+	CHERITEST_VERIFY2(c2 == NULL, "Initial read NULL");
+
+	mmap_and_check_tag_stored(fd, PROT_READ | PROT_WRITE, MAP_SHARED);
+
+	/* And now verify that it is, thanks to the aliased maps */
+	c2 = *map2;
+	CHERITEST_VERIFY2(cheri_gettag(c2) != 0, "tag lost on second read");
+	CHERITEST_VERIFY2(c2 != NULL, "Second read not NULL");
+
+	cheritest_success();
+}
+
+void
+cheritest_vm_shm_open_anon_unix_surprise(const struct cheri_test *ctp __unused)
+{
+	int sv[2];
+	int pid;
+
+	CHERITEST_CHECK_SYSCALL(socketpair(AF_UNIX, SOCK_DGRAM, 0, sv) != 0);
+
+	pid = fork();
+	if (pid == -1)
+		cheritest_failure_errx("Fork failed; errno=%d", errno);
+
+	if (pid == 0) {
+		void * __capability * map;
+		void * __capability c;
+		int fd;
+		struct msghdr msg = { 0 };
+		struct cmsghdr * cmsg;
+		char cmsgbuf[CMSG_SPACE(sizeof(fd))] = { 0 } ;
+		char iovbuf[16];
+		struct iovec iov = { .iov_base = iovbuf, .iov_len = sizeof(iovbuf) };
+
+		close(sv[1]);
+
+		/* Read from socket */
+		msg.msg_iov = &iov;
+		msg.msg_iovlen = 1;
+		msg.msg_control = cmsgbuf;
+		msg.msg_controllen = sizeof(cmsgbuf);
+		CHERITEST_CHECK_SYSCALL(recvmsg(sv[0], &msg, 0));
+
+		/* Deconstruct cmsg */
+		/* XXX Doesn't compile: cmsg = CMSG_FIRSTHDR(&msg); */
```

Comment:
Is this a clang bug or one in the definition of `CMSG_FIRSTHDR`?

Reply:
The latter, I think, with some missing distinction between about purecap vs. hybrid worlds?  The error in particular is

```
/mnt/mainline/cheribsd/bin/cheritest/cheritest_vm.c:188:10: error: invalid source type 'void *' for __cheri_fromcap: source must be a capability
                cmsg = CMSG_FIRSTHDR(&msg);
                       ^
/junk/cheri-build/cheribsd-obj-128/mnt/mainline/cheribsd/mips.mips64/tmp/usr/include/sys/socket.h:585:37: note: expanded from macro 'CMSG_FIRSTHDR'
         (__cheri_fromcap struct cmsghdr *)(mhdr)->msg_control : \
                                           ^
1 error generated.
```